File: test/language/parameterized_types/cpp/ParameterizedParamTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "parameterized_types/parameterized_param/ParameterizedParamHolder.h"

namespace parameterized_types
{
namespace parameterized_param
{

class ParameterizedParamTest : public ::testing::Test
{
protected:
    void fillParameterizedParamHolder(ParameterizedParamHolder& parameterizedParamHolder)
    {
        Param& param = parameterizedParamHolder.getParam();
        param.setValue(PARAM_VALUE);
        param.setExtraValue(PARAM_EXTRA_VALUE);

        ParameterizedParam& parameterizedParam = parameterizedParamHolder.getParameterizedParam();
        parameterizedParam.setValue(PARAMETERIZED_PARAM_VALUE);
        parameterizedParam.setExtraValue(PARAMETERIZED_PARAM_EXTRA_VALUE);

        parameterizedParamHolder.initializeChildren();
    }

    void checkParameterizedParamHolderInBitStream(zserio::BitStreamReader& reader,
            const ParameterizedParamHolder& parameterizedParamHolder)
    {
        ASSERT_EQ(parameterizedParamHolder.getParameter(), reader.readBits(16));

        const Param& param = parameterizedParamHolder.getParam();
        ASSERT_EQ(param.getParameter(), PARAMETER);
        ASSERT_EQ(param.getValue(), reader.readBits(16));
        ASSERT_EQ(param.getExtraValue(), reader.readBits(32));

        const ParameterizedParam& parameterizedParam = parameterizedParamHolder.getParameterizedParam();
        ASSERT_EQ(parameterizedParam.getParam(), param);
        ASSERT_EQ(parameterizedParam.getValue(), reader.readBits(16));
        ASSERT_EQ(parameterizedParam.getExtraValue(), reader.readBits(32));
    }

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);

private:
    static const uint16_t PARAMETER;
    static const uint16_t PARAM_VALUE;
    static const uint32_t PARAM_EXTRA_VALUE;
    static const uint16_t PARAMETERIZED_PARAM_VALUE;
    static const uint32_t PARAMETERIZED_PARAM_EXTRA_VALUE;
};

const uint16_t ParameterizedParamTest::PARAMETER = 11;
const uint16_t ParameterizedParamTest::PARAM_VALUE = 0x0BED;
const uint32_t ParameterizedParamTest::PARAM_EXTRA_VALUE = 0x0BEDDEAD;
const uint16_t ParameterizedParamTest::PARAMETERIZED_PARAM_VALUE = 0x0BAD;
const uint32_t ParameterizedParamTest::PARAMETERIZED_PARAM_EXTRA_VALUE = 0x0BADDEAD;

TEST_F(ParameterizedParamTest, write)
{
    ParameterizedParamHolder parameterizedParamHolder;
    fillParameterizedParamHolder(parameterizedParamHolder);

    zserio::BitStreamWriter writer(bitBuffer);
    parameterizedParamHolder.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkParameterizedParamHolderInBitStream(reader, parameterizedParamHolder);
    reader.setBitPosition(0);

    ParameterizedParamHolder readParameterizedParamHolder(reader);
    ASSERT_EQ(parameterizedParamHolder, readParameterizedParamHolder);
}

} // namespace parameterized_param
} // namespace parameterized_types

File: test/extensions/language/parameterized_types/python/ParameterizedParamTest.py
import zserio

import ParameterizedTypes


class ParameterizedParamTest(ParameterizedTypes.TestCase):
    def testWrite(self):
        parameterizedParamHolder = self._createParameterizedParamHolder()
        writer = zserio.BitStreamWriter()
        parameterizedParamHolder.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        self._checkParameterizedParamHolderInStream(reader, parameterizedParamHolder)

        reader.bitposition = 0
        readParameterizedParamHolder = self.api.ParameterizedParamHolder.from_reader(reader)
        self.assertEqual(parameterizedParamHolder, readParameterizedParamHolder)

    def _createParameterizedParamHolder(self):
        param = self.api.Param(self.PARAMETER, self.PARAM_VALUE, self.PARAM_EXTRA_VALUE)
        parameterizedParam = self.api.ParameterizedParam(
            param, self.PARAMETERIZED_PARAM_VALUE, self.PARAMETERIZED_PARAM_EXTRA_VALUE
        )

        return self.api.ParameterizedParamHolder(self.PARAMETER, param, parameterizedParam)

    def _checkParameterizedParamHolderInStream(self, stream, parameterizedParamHolder):
        self.assertEqual(parameterizedParamHolder.parameter, stream.read_bits(16))

        param = parameterizedParamHolder.param
        self.assertEqual(param.parameter, self.PARAMETER)
        self.assertEqual(param.value, stream.read_bits(16))
        self.assertEqual(param.extra_value, stream.read_bits(32))

        parameterizedParam = parameterizedParamHolder.parameterized_param
        self.assertEqual(parameterizedParam.param, param)
        self.assertEqual(parameterizedParam.value, stream.read_bits(16))
        self.assertEqual(parameterizedParam.extra_value, stream.read_bits(32))

    PARAMETER = 11
    PARAM_VALUE = 0x0BED
    PARAM_EXTRA_VALUE = 0x0BEDDEAD
    PARAMETERIZED_PARAM_VALUE = 0x0BAD
    PARAMETERIZED_PARAM_EXTRA_VALUE = 0x0BADDEAD
